### PR TITLE
Add a method `.major()` for returning the major version number

### PR DESCRIPTION
**Pull request by LucianBuzzo**
Connects to #13

change-type: minor

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -109,3 +109,49 @@ export const compare = memoize((versionA: string | null, versionB: string | null
 	}
 	return versionA.localeCompare(versionB);
 }, (a: string, b: string) => `${a} && ${b}`);
+
+/**
+ * @summary Return the major version number
+ * @name compare
```
Comment (thgreasi) [inline src/index.ts]: I guess that this should say `@name major` and everything else will change accordingly.

File: src/index.ts [hunk 2]
```diff
@@ -109,3 +109,49 @@ export const compare = memoize((versionA: string | null, versionB: string | null
...
+ * @example
+ * resinSemver.major(null); //--> 0
+ *
+ * resinSemver.major('Resin OS v2.0.5'); //--> 2
+ *
+ * resinSemver.major('Resin OS v2.0.5'); //--> 2
+ *
+ * resinSemver.major('Linux 14.04'); //--> 14
+ *
+ * resinSemver.major('My development version'); //--> 1
```
Comment (pimterry) [inline src/index.ts]: This logic is where we start getting inconsistencies. It's debatable how they should work out, but we should have some coherent answers. Intuitively, I want to assume that if compare says A > B, then major(A) >= major(B) should be true. I.e. if one version is newer than another, then I should be able to assume it's the same or a later major version.

If we want to make that assumption hold, then these examples don't work (v0.0.1 compares above My development version, but has a lower major result).

The same holds for strings that happen to have numbers in them (Linux 14.04): we should treat them similarly across all the methods here. That means if we're going to optimistically try to infer the version number from them here, then we should use that same inferred number in compare.

Totally happy to debate what the rules for this to be, but they should be consistent.

Personally: I'd do proper semver comparison and major/minor/etc parsing only on valid semver strings plus a couple of clear specific semver-compatible equivalents (e.g. dropping Resin OS v prefixes, fixing specific not-quite-semver suffixes like .dev), and for everything else compare them below, and return null for major/minor/etc. Essentially we do comparison on the set of versions we're really confident about, and we treat everything else as equivalent 'other'. How does that sound?

File: test/index.spec.ts [hunk 3]
```diff
@@ -12,4 +12,29 @@ describe('resin-semver', () => {
 			});
 		});
 	});
+
+	describe('.major()', () => {
+		it('should correctly match valid semver versions', () => {
+			expect(semver.major('4.05.1')).to.equal(4);
```
Comment (pimterry) [inline test/index.spec.ts]: This isn't actually a valid semver version (`semver.valid('4.05.1')` returns null)

File: src/index.ts (hunk 2 above)
Comment (LucianBuzzo) [inline src/index.ts]: In summary.
- valid semver and "normalized" semver (for Resin OS prefixes) are compared in the normal way.
- sermverish strings will always be weighted above everything else
- comparing to non-semverish strings will devolve to alphabetical sort
- extraction methods (major/minor/etc) will return null for non-semverish strings

Comment (pimterry) [inline src/index.ts]: What does semverish mean?

Comment (LucianBuzzo) [inline src/index.ts]: by semverish, I mean a string that is either a valid semver, or a semver string that has the resin prefixes and suffixes, EG "ResinOS 2.0.2+rev2 (prod)".

Comment (pimterry) [inline src/index.ts]: Ahhhh, ok, gotcha. In that case this sounds great to me :+1:

Comment (LucianBuzzo) [inline src/index.ts]: I think I'll add a cover note explaining this approach in the README

File: test/index.spec.ts [hunk 4]
```diff
@@ -12,4 +12,26 @@ describe('resin-semver', () => {
...
+			expect(semver.major('Resin OS 2.0.2+rev2')).to.equal(2);
+			expect(semver.major('Resin OS 2.0.0.rev1 (prod)')).to.equal(2);
+			expect(semver.major('Resin OS 2.0.0-rc5.rev1')).to.equal(2);
+		});
+
+		it('should return null when version is `null`', () => {
+			expect(semver.major(null)).to.equal(null);
+		});
+
+		it('should return null when the version is contains no valid semver value', () => {
```
Comment (pimterry) [inline test/index.spec.ts]: 'version is contains'

File: test/index.spec.ts [hunk 5]
```diff
@@ -12,4 +12,26 @@ describe('resin-semver', () => {
...
+			expect(semver.major('My dev version')).to.equal(null);
+			expect(semver.major('Linux 14.04')).to.equal(null);
```
Comment (pimterry) [inline test/index.spec.ts]: Can we add a case for non-ResinOS prefixed string that contains a semver 3-part version somewhere, just to make it clear (and guarantee for the future), that we generally don't parse versions like that?

As a fun example (and a great explanation of why comparing other people's versions is scary) a little research earlier came up with OpenSuse Leap's versioning scheme. They're currently on version `42.3.20170726.72bbcf8`, [to be followed next year](https://www.phoronix.com/scan.php?page=news_item&px=openSUSE-Leap-15-Versioning) by `15.x.x`, of course. :sob:

**Pull request merged**